File: src/main/java/cz/vutbr/fit/interlockSim/gui/StatusBar.java
/* Brno University of Technology
 * Faculty of Information Technology
 *
 * BSc Thesis  2006/2007
 *
 * Railway Interlocking Simulator
 *
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.gui;

import java.awt.Component;
import java.awt.Dimension;
import java.awt.event.MouseEvent;
import java.awt.event.MouseMotionListener;
import java.beans.PropertyChangeEvent;

import javax.swing.JLabel;

import cz.vutbr.fit.interlockSim.Main;
import cz.vutbr.fit.interlockSim.context.ContextChangeListener;

/**
 * Collect and show status information
 *
 */
public class StatusBar extends JLabel implements ContextChangeListener {
	private class MouseL implements MouseMotionListener {
		public void mouseDragged(MouseEvent e) {
			// EMPTY
		}

		public void mouseMoved(MouseEvent e) {
			Object source = e.getSource();
			assert source instanceof StatusProducer;
			final String status = ((StatusProducer) source).getStatus(e);
			if (status != null) setText(status);
		}

	}

	private final MouseL mouseL = new MouseL();

	public StatusBar() {
		setPreferredSize(new Dimension(100, 25));
		setText("Welcome to " + Main.PROGRAM_NAME);
	}

	private Component checkComponent(StatusProducer producer) {
		assert producer instanceof Component : producer;
		Component producerComponent = (Component) producer;
		return producerComponent;
	}

	public void registerProducer(StatusProducer producer) {
		Component producerComponent = checkComponent(producer);
		producerComponent.addMouseMotionListener(mouseL);
	}

	public void unregisterProducer(StatusProducer producer) {
		Component producerComponent = checkComponent(producer);
		producerComponent.removeMouseMotionListener(mouseL);
	}

	@Override
	public void propertyChange(PropertyChangeEvent evt) {
		Object newValue = evt.getNewValue();
		if (newValue instanceof CharSequence) {
			setText(newValue.toString());
		} else if (newValue != null) {
			setText(newValue.toString());
		}
	}
}

File: src/main/kotlin/cz/vutbr/fit/interlockSim/gui/StatusBar.kt
/* Brno University of Technology
 * Faculty of Information Technology
 *
 * BSc Thesis  2006/2007
 *
 * Railway Interlocking Simulator
 *
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.gui

import cz.vutbr.fit.interlockSim.PROGRAM_NAME
import cz.vutbr.fit.interlockSim.context.ContextChangeListener
import cz.vutbr.fit.interlockSim.exceptions.requireValidState
import cz.vutbr.fit.interlockSim.objects.cells.ContextChangeEvent
import java.awt.Component
import java.awt.Dimension
import java.awt.event.MouseEvent
import java.awt.event.MouseMotionListener
import javax.swing.JLabel

/**
 * Status bar for displaying context information and mouse motion status
 */
class StatusBar :
	JLabel(),
	ContextChangeListener {
	private val mouseListener =
		object : MouseMotionListener {
			override fun mouseDragged(e: MouseEvent) {
				// Not used
			}

			override fun mouseMoved(e: MouseEvent) {
				val source = e.source
				requireValidState(source is StatusProducer) { "Source must be a StatusProducer" }
				val status = (source as StatusProducer).getStatus(e)
				text = status
			}
		}

	init {
		preferredSize = Dimension(100, 25)
		text = "Welcome to " + PROGRAM_NAME
	}

	private fun checkComponent(producer: StatusProducer): Component {
		requireValidState(producer is Component) { "StatusProducer must be a Component" }
		return producer as Component
	}

	fun registerProducer(producer: StatusProducer) {
		val producerComponent = checkComponent(producer)
		producerComponent.addMouseMotionListener(mouseListener)
	}

	fun unregisterProducer(producer: StatusProducer) {
		val producerComponent = checkComponent(producer)
		producerComponent.removeMouseMotionListener(mouseListener)
	}

	override fun propertyChange(event: ContextChangeEvent) {
		val newValue = event.newValue
		when {
			newValue is CharSequence -> text = newValue.toString()
			newValue != null -> text = newValue.toString()
		}
	}

	/**
	 * Shows a temporary message in the status bar.
	 * The message will be displayed for the specified duration, then cleared.
	 *
	 * @param message The message to display
	 * @param durationMs Duration in milliseconds (default: 3000ms = 3 seconds)
	 */
	fun showTemporaryMessage(
		message: String,
		durationMs: Long = 3000
	) {
		val originalText = text
		text = message

		// Use Swing Timer to restore original text after duration
		val timer =
			javax.swing.Timer(durationMs.toInt()) {
				text = originalText
			}
		timer.isRepeats = false
		timer.start()
	}
}
